vortex-gpu/src/rle_decompress.rs

```diff
@@ -0,0 +1,131 @@
+                rle.indices().to_primitive().as_slice::<u16>(),
+                rle.values().to_primitive().as_slice::<V>(),
+                ctx,
+            ),
+            _ => vortex_panic!(
+                "Unsupported index type for RLE decoding: {}",
+                rle.indices().dtype().as_ptype()
+            ),
+        }
+    })
+}
+
+fn cuda_rle_decompress_typed<Values, Indices>(
+    indices: &[Indices],
+    values: &[Values],
+    ctx: Arc<CudaContext>,
+) -> VortexResult<ArrayRef>
+where
+    Values: NativePType + DeviceRepr + ValidAsZeroBits,
+    Indices: UnsignedPType + DeviceRepr,
+{
+    assert_eq!(indices.len() % 1024, 0);
+
+    let kernel_func = cuda_take_kernel::<Indices, Values>(false, ctx.clone())?;
+    let num_chunks =
+        u32::try_from(indices.len().div_ceil(1024)).vortex_expect("num chunks overflow");
+    let stream = ctx.default_stream();
+
+    let cu_values = stream
+        .memcpy_stod(values)
+        .map_err(|e| vortex_err!("Failed to copy to device: {e}"))?;
+    let cu_indices = stream
+        .memcpy_stod(indices)
+        .map_err(|e| vortex_err!("Failed to copy to device: {e}"))?;
+
+    let mut cu_out = unsafe {
+        stream
+            .alloc::<Values>(indices.len().next_multiple_of(1024))
+            .map_err(|e| vortex_err!("Failed to allocate stream: {e}"))?
+    };
+
+    let mut launch = stream.launch_builder(&kernel_func);
+    launch.arg(&cu_indices);
+    launch.arg(&cu_values);
+    launch.arg(&mut cu_out);
+    unsafe {
+        launch.launch(LaunchConfig {
+            grid_dim: (num_chunks, 1, 1),
+            block_dim: (32, 1, 1),
+            shared_mem_bytes: 0,
+        })
+    }
+    .map_err(|e| vortex_err!("Failed to launch: {e}"))?;
+
+    let mut buffer = BufferMut::<Values>::with_capacity(indices.len());
+    unsafe { buffer.set_len(indices.len()) }
+
+    stream
+        .memcpy_dtoh(&cu_out, &mut buffer)
+        .map_err(|e| vortex_err!("Failed to copy to device: {e}"))?;
+    stream
+        .synchronize()
+        .map_err(|e| vortex_err!("Failed to synchronize: {e}"))?;
+
+    Ok(PrimitiveArray::new(buffer, Validity::NonNullable).into_array())
+}
+
+#[cfg(all(target_os = "linux", feature = "cuda"))]
+#[cfg(test)]
+mod tests {
+    use std::iter;
+
+    use cudarc::driver::CudaContext;
+    use vortex_array::arrays::PrimitiveArray;
+    use vortex_array::validity::Validity;
+    use vortex_array::ToCanonical;
+    use vortex_buffer::Buffer;
+    use vortex_error::VortexUnwrap;
+    use vortex_fastlanes::RLEArray;
+
+    use crate::rle_decompress::cuda_rle_decompress;
+
+    #[test]
+    fn test_cuda_rle_decompress() {
+        let primitive_array = PrimitiveArray::new(
+            (0u32..4096)
+                .flat_map(|i| iter::repeat_n(i, i as usize % 10))
+                .collect::<Buffer<_>>(),
+            Validity::NonNullable,
+        );
+        let array = RLEArray::encode(&primitive_array).vortex_unwrap();
```

Comment:
You need to factor in that indices in RLEArray are chunk local, as in relative to the chunk start. So its a dictionary lookup within the 1024 slice of the chunk.